fix(scheduling_queue): always put Pods with no unschedulable plugins into activeQ/backoffQ

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR changes how the scheduling queue regards the scheduling failure.
We sometimes put the Pod back to the queue with the unschedulable plugins, and sometimes we don't.
The difference between these two cases is that the former is the rejection from the plugin, and the latter is the error from the plugin (or the scheduler, which shouldn't happen unless the scheduler has bugs).
More specifically, when the Pod is rejected in PreFilter, Filter, Reserve or Permit, the scheduler attached the failed plugin name to the Pod and put it back to the queue. And, when the Pod is rejected in another extension point (like Bind etc) which is probably because of the kube-apiserver failure (or the bug in the plugin impl), the scheduler put the pod to the queue without any plugin names.
In cases of the latter, it's more likely unexpected rejection and maybe it won't be resolved by cluster events.
This PR changes the scheduling queue to always put Pods without the failed plugins into activeQ/backoffQ so that we can prevent such Pods stuck in unschedQ forever.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref: https://github.com/kubernetes/kubernetes/pull/118438#discussion_r1252301822

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
More accurate requeueing in scheduling queue for Pods rejected by the temporal failure (e.g., temporal failure on kube-apiserver.)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
Are we able to reproduce the potential bug, and then verify the fix does work?

sanposhiho replied:
Sorry for the delay, I'll work on that this weekend.

sanposhiho replied:
I added TestRequeueByPermitRejection. 
Please check out https://github.com/sanposhiho/kubernetes/commit/e0cd7a09a9022b53af74f13bc71ea69a0881b80f.
You can see a failed test result where I run the same test in the master branch.

sanposhiho commented:
@alculquicondor I think that's a good point and could be another blocker of the removal of the flush because the Pod would stuck in unschedQ in a worst case if we didn't have a flush. 

But, I'd say that's kind of difficult from the scheduler side to know whether the error is coming from the kube-apiserver unstable or not.. Because, in the first place, the scheduler cannot know which plugin in which extension point communicates with kube-apiserver. 
What do you think? Do we need to create some special returning status so that plugin can let the scheduler know "this failure is coming from kube-apiserver and the queue should requeue the pod into activeQ/backoffQ in any ways"?

alculquicondor commented:
> But, I'd say that's kind of difficult from the scheduler side to know whether the error is coming from the kube-apiserver unstable or not.

I don't think we need to make it specific to apiserver errors, but errors in general (as opposed to "Unschedulable(AndUnResolvable)" statuses).

sanposhiho commented:
> errors in general (as opposed to "Unschedulable(AndUnResolvable)" statuses).

Sounds good. Let me have an issue for this.

alculquicondor commented:
why not fix in this PR, as it's highly related?

sanposhiho commented:
I added TestRequeueByPermitRejection that reproduce a bug in the master branch and pass in this branch. 
Also, I fixed [this issue](https://github.com/kubernetes/kubernetes/pull/119105#issuecomment-1629201087) as well in this branch. (see the PR description at the top, as I updated)

alculquicondor commented:
I guess we can abandon this PR?

Huang-Wei commented:
> I guess we can abandon this PR?

We can merge #118438 and then use this PR to only add the integration test https://github.com/kubernetes/kubernetes/pull/119105#discussion_r1265588646. I verified the integration test passed atop #118438, and failed on master.

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -595,7 +582,13 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 		metrics.UnschedulableReason(plugin, pInfo.Pod.Spec.SchedulerName).Inc()
 	}
-	if p.moveRequestCycle >= podSchedulingCycle {
+	if p.moveRequestCycle >= pInfo.SchedulingCycle || len(pInfo.UnschedulablePlugins) == 0 {
```
Also, this change is still worthwhile for the kube-apiserver failure ([Aldo's comment](https://github.com/kubernetes/kubernetes/pull/119105#issuecomment-1629201087)).
When len(pInfo.UnschedulablePlugins) == 0, it assumes that some unexpected failure happened and put the Pod back to the activeQ/backoffQ. 

I'll rebase + revert unnecesarry changes later!

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -201,9 +201,6 @@ type PriorityQueue struct {
 	metricsRecorder metrics.MetricAsyncRecorder
 	// pluginMetricsSamplePercent is the percentage of plugin metrics to be sampled.
 	pluginMetricsSamplePercent int
-
-	// isSchedulingQueueHintEnabled indicates whether the feature gate for the scheduling queue is enabled.
-	isSchedulingQueueHintEnabled bool
```
follow up: https://github.com/kubernetes/kubernetes/pull/118438#discussion_r1265806317

alculquicondor replied:
I think my comment was misinterpreted.

It is ok to have a proxy boolean for the feature gate (in fact, that's the only way to write unit-tests that can run in parallel).

What is not ok is not to guard a piece of code with a feature gate check based on the assumption that another section is not buggy.

sanposhiho replied:
Understood. 🙏

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -606,14 +602,12 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
 
 // determineSchedulingHintForInFlightPod looks at the unschedulable plugins of the given Pod
 // and determines the scheduling hint for this Pod while checking the events that happened during in-flight.
-func (p *PriorityQueue) determineSchedulingHintForInFlightPod(logger klog.Logger, pInfo *framework.QueuedPodInfo, podSchedulingCycle int64) framework.QueueingHint {
```
Sorry, podSchedulingCycle is an unused parameter. 🙏

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -439,3 +439,111 @@ func TestCustomResourceEnqueue(t *testing.T) {
+
+	node.Labels = map[string]string{"updated": ""}
+	if _, err := cs.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("Failed to remove taints off the node: %v", err)
```
```suggestion
		t.Fatalf("Failed to add labels to the node: %v", err)
```

sanposhiho commented on test/integration/scheduler/queue_test.go:
```diff
@@ -439,3 +439,111 @@ func TestCustomResourceEnqueue(t *testing.T) {
+	}
+
+	// reject pod-1 to simulate the failure in Permit plugins.
+	// This pod-1 should be enqueued to activeQ because the NodeUpdate event has happened.
```
> A failed Permit plugin doesn't get added to pInfo.UnschedulablePlugins?

Talking about the current impl, it does in the Permit extension, but doesn't in WaitOnPermit. So, failed in WaitOnPermit is handled in the same way as other extension points in binding cycle.
But, agree to register in `UnschedulablePlugins` in WaitOnPermit as well. I'll have another PR then.

sanposhiho replied:
https://github.com/kubernetes/kubernetes/pull/119785

sanposhiho commented on test/integration/scheduler/queue_test.go:
```diff
@@ -439,3 +439,111 @@ func TestCustomResourceEnqueue(t *testing.T) {
+
+// TestRequeueByPermitRejection verify Pods failed by permit plugins in the binding cycle are
+// put back to the queue, according to the correct scheduling cycle number.
+func TestRequeueByPermitRejection(t *testing.T) {
```
Sure, added.

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -732,6 +736,9 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 	// In this case, we try to requeue this Pod to activeQ/backoffQ.
 	queue := p.requeuePodViaQueueingHint(logger, pInfo, schedulingHint, ScheduleAttemptFailure)
 	logger.V(5).Info("Pod moved to an internal scheduling queue", "pod", klog.KObj(pod), "event", ScheduleAttemptFailure, "queue", queue, "schedulingCycle", podSchedulingCycle)
+	if queue == activeQ {
+		p.cond.Broadcast()
+	}
```
TestRequeueByBindFailure doesn't pass without this change.
It's a bug and I have another PR to fix this asap:
https://github.com/kubernetes/kubernetes/pull/119784

alculquicondor reviewed (commented):
/lgtm

We should put this in 1.28.1

sanposhiho commented:
> We should put this in 1.28.1

Sure, will do after I'm back to the normal days - this Friday.

/unhold

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -726,16 +726,20 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 	}
+=======
+	logger.V(5).Info("Pod moved to an internal scheduling queue", "pod", klog.KObj(pod), "event", ScheduleAttemptFailure, "queue", queue, "schedulingCycle", podSchedulingCycle)
+>>>>>>> dc313b9f0b9 (always put Pods with no unschedulable plugins into activeQ/backoffQ)
```
Please resolve this.

sanposhiho replied:
Oh, sorry, that's already fixed in the latest commit. I'll squash commits when doing rebase. 🙏

alculquicondor commented:
oh, this is the PR I was thinking of yesterday, @pohly

sanposhiho committed:
always put Pods with no unschedulable plugins into activeQ/backoffQ

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -3180,3 +3203,11 @@ func Test_isPodWorthRequeuing(t *testing.T) {
 	}
 }
+
+func mustAddUnschedulableIfNotPresent(t *testing.T, q *PriorityQueue, logger klog.Logger, pInfo *framework.QueuedPodInfo, podSchedulingCycle int64) {
```
To avoid ↓ in all places.

```
ERROR: pkg/scheduler/internal/queue/scheduling_queue_test.go:1559:32: Error return value of `q.AddUnschedulableIfNotPresent` is not checked (errcheck)
ERROR: 	q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(highPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119105/pull-kubernetes-linter-hints/1700123530751905792

sanposhiho commented:
@Huang-Wei @alculquicondor rebase done. 🙏

alculquicondor commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -1113,8 +1112,7 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 
-	// (no need to check the feature gate because there is always no p.inFlightPods when the feature is disabled.)
-	if len(p.inFlightPods) != 0 {
+	if len(p.inFlightPods) != 0 && p.isSchedulingQueueHintEnabled {
```
for readability, I would swap the order of these two checks.

alculquicondor commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -1325,31 +1333,44 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
 	if p, err := q.Pop(); err != nil || p.Pod != hpp1 {
 		t.Errorf("Expected: %v after Pop, but got: %v", hpp1, p.Pod.Name)
 	}
+	unschedulableQueuedPodInfo := q.newQueuedPodInfo(unschedulablePodInfo.Pod, "fooPlugin")
```
This entire test is too hard to read (anyone reading has to keep a mental state of all that happened in the previous steps). I wonder if it can be split into different test cases that target specific transitions.
But that would be for a follow up.

sanposhiho replied:
👍 
I'd like to do that refactor in a follow-up PR. 
I guess somehow we can just merge this test and `TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint`.

sanposhiho committed:
address review comments

sanposhiho commented:
@alculquicondor fixed in https://github.com/kubernetes/kubernetes/pull/119105/commits/9932089472eb15b383ddd5355cfdda26529deab9. 
Let me know if it's ok to squash them
.

alculquicondor commented:
/lgtm
/label tide/merge-method-squash
